calico/felix/config.py

```diff
@@ -46,166 +46,311 @@
+        except KeyError:
+            raise ConfigException("Invalid log level",
+                                  self.parameters["LogSeveritySys"])
+
+        try:
+            self.LOGLEVSCR = LOGLEVELS[self.LOGLEVSCR.lower()]
+        except KeyError:
+            raise ConfigException("Invalid log level",
+                                  self.parameters["LogSeverityScreen"])
+
+        # Log file may be "None" (the literal string, case insensitive). In
+        # this case no log file should be written.
+        if self.LOGFILE.lower() == "none":
+            self.LOGFILE = None
 
-    def validate_cfg(self):
-        #*********************************************************************#
-        #* Firewall that the config is not invalid.                          *#
-        #*********************************************************************#
         if self.METADATA_IP.lower() == "none":
             # Metadata is not required.
             self.METADATA_IP = None
             self.METADATA_PORT = None
         else:
             # Metadata must be supplied as IP or address, but we store as IP
-            self.METADATA_IP = self.validate_addr("MetadataAddr",
+            self.METADATA_IP = self._validate_addr("MetadataAddr",
                                                   self.METADATA_IP)
 
             if not common.validate_port(self.METADATA_PORT):
-                raise ConfigException("Invalid MetadataPort value : %s" %
-                                      self.METADATA_PORT,
-                                      "etcd:/calico/config/MetadataPort")
-
-
-        if self.IFACE_PREFIX is None:
-            raise ConfigException("Missing InterfacePrefix value",
-                                  "etcd:/calico/config/InterfacePrefix")
-
-        # Log file may be "None" (the literal string, either provided or as
-        # default). In this case no log file should be written.
-        if self.LOGFILE.lower() == "none":
-            self.LOGFILE = None
-
-    def warn_unused_cfg(self, cfg_dict):
-        # Firewall that no unexpected items in the config file - i.e. ones we
-        # have not used.
-        for section in self._items.keys():
-            for lKey in self._items[section].keys():
-                log.warning("Got unexpected item %s=%s in %s",
-                            lKey, self._items[section][lKey], self._config_path)
-
+                raise ConfigException("Invalid field value",
+                                      self.parameters["MetadataPort"])
+
+        if not final:
+            # Do not check that unset parameters are defaulted; we have more
+            # config to read.
+            return
+
+        for name, parameter in self.parameters.iteritems():
+            if parameter.value is None:
+                # No value, not even a default
+                raise ConfigException("Missing undefaulted value",
+                                      self.parameters["InterfacePrefix"])
```

Comment:
Looks like the error is specific to InterfacePrefix even though it's in a generic loop?